src/main/java/com/conveyal/r5/analyst/DataPreloader.java

```diff
@@ -0,0 +1,239 @@
+ * trigger the distance table computation, then a third one to link the destination grids for a particular mode.
+ * We want all the steps to be done at one go. Therefore we want a single asynchronous action at the top level, which
+ * cascades a lot of sequential synchronous actions interspersed with progress updates.
+ *
+ * The distance table construction is done as part of the scenario application. So our main steps are:
+ * fetch the network, apply the scenario, build the linkages.
+ *
+ * The incoming request may not have a scenario ID in it. TransportNetwork.getNetworkForScenario() resolves it as
+ * follows: String scenarioId = request.scenarioId != null ? request.scenarioId : request.scenario.id;
+ *
+ * I was initially trying to do this with a Caffeine AsyncLoadingCache, but ran into the following problems:
+ * 1. The cache values must be computed from the key only. But expanded scenarios may be contained in the request.
+ * 2. We want to provide progress information, and retrofitting the Caffeine cache with it would be just as complicated
+ *    as making our own new class.
+ * 3. We don't really want to trigger actions after the asynchronous loading is finished, we just want to know
+ *    immediately at the time of each call whether it's finished building or not.
+ *
+ * The steps that need to happen are the following:
+ * 1. Download and load the network if it's already built.
+ * 2. Build the network if it's not already built.
+ * 3. Apply the scenario to the network, or fetch the resulting network if the scenario is already applied.
+ * 3. Pre-compute stop-to-vertex distance tables and linkages (stop-to-point distance tables) for the walking mode.
+ * 4. Create or fetch a grid of destinations based on this specific task and the gridCache.
+ * 5. Link that grid to the street network with the right mode, basing distances on the stop-to-vertex tables created above.
+ *
+ * Repeatedly retrying a full task just to see if it fails seems like a bad way to report progress.
+ * We may want to enqueue tasks like we do in data tools, with progress reporting. That's the only thing that's going
+ * to give people a sense that it's good enough.
+ *
+ * However that's the first step to getting this solved - the nested progress tracking and reporting part can come later.
+ *
+ * A tricky thing is how to handle the question: is it available now or not? If it is available,
+ * Maybe you do want to kick off the process, block for half a second waiting on the result, then report where you're at.
+ *
+ * Perhaps the values should be a wrapper type bundling together all the data loaded all at once by the preloader.
+ * Our current design is somewhat suboptimal in that all of these items are in different caches.
+ * The distance tables or linkages could be evicted for example.
+ *
+ * Created by abyrd on 2018-09-17
+ */
+public class DataPreloader extends AsyncLoader<DataPreloader.Key, TransportNetwork> {
```

Comment:
Good point. I renamed it to `NetworkPreloader`, in 48390b152d424460e3cb4df2fdf7770feed0afab because it subclasses `AsyncLoader` (loader here referencing the idea of a LoadingCache) and I find it useful to remember that its purpose is to load in advance of any analysis happening (pre-).

This is still not ideal because a lot of the time is not spent on the network (but rather on the linkages and tables). But it's already a bit less generic.